src/app/application/services/data-source/DataSource.ts

```diff
@@ -0,0 +1,83 @@
+import CallableEventSource, { isClosed } from './CallableEventSource'
+import Router from './Router'
+import SharedPool from './SharedPool'
+
+export type DataSourceResponse<T> = {data: T | undefined, error: Error | undefined, refresh: () => void}
+export type Source = (params: Record<string, unknown>, source: {close: () => void}) => Promise<unknown>
+export type Sources = Record<string, Source>
+
+const get = (src: string, router: Router<Source>) => {
+  const [path, query] = src.split('?')
+  const queryParams = new URLSearchParams(query)
+  const route = router.match(path)
+  const _source = new CallableEventSource(async function * (this: CallableEventSource) {
+    while (true) {
+      this.readyState = 1
+      // latency
+      // await new Promise(resolve => setTimeout(resolve, 5000))
+      yield route.route({
+        ...{
+          offset: parseInt(queryParams.get('offset') || '0'),
+        },
+        ...route.params,
+      }, this)
+      if (!isClosed(this)) {
+        await new Promise(resolve => setTimeout(resolve, 5000))
+      }
+    }
+  })
+  return _source as EventSource
+}
+const destroy = (_src: string, source: EventSource) => {
+  if (source) {
+    source.close()
+  }
+}
+
+export class DataSource {
+  cache: Map<string, unknown> = new Map()
+  pool: SharedPool<string, EventSource>
+  constructor(requests: Sources) {
+    const requestRouter: Router<Source> = new Router(requests)
+
+    this.pool = new SharedPool<string, EventSource>(
+      (src: string) => {
+        return get(src, requestRouter)
+      },
+      (src: string, source: EventSource) => {
+        return destroy(src, source)
+      },
+    )
+  }
+
+  source(src: string, ref: symbol) {
+    return this.open(src, ref)
+  }
+
+  cached(src: string) {
+    return this.cache.get(src)
+  }
+
+  open(src: string, ref: symbol): EventSource | undefined {
+    const _source = this.pool.acquire(src, ref)
+    if (typeof _source !== 'undefined') {
+      // _source.open();
+      _source.addEventListener('message', (e: Event) => {
+        // always fill the cache on a successful response
+        this.cache.set(src, (e as MessageEvent).data)
+      })
```

Comment:
**Suggestion**: Functions defined directly at the callsite of a properly typed function should typically not have or need type annotations. In this case, typing the `event` of the listener as `Error` prevents TypeScript from inferring it as `MessageEvent`.

```suggestion
      _source.addEventListener('message', (e) => {
        // always fill the cache on a successful response
        this.cache.set(src, e.data)
      })
```

Reply:
Ah I think this was from an earlier noodle, seeing as I typed everything to native EventSource pretty late on. I've also noticed that this is required for the firing and ErrorEvent as it seems like EventSource doesn't fire those. I guess probably because ErrorEvents didn't exist when EventSource started out maybe 🤔 

Anyway, whilst this seems like an easy fix, I want to think about it first. Considering this has very little impact I believe, and is required in the current form the case of ErrorEvent I'm not too concerned rn. I might change the MessageEvent one but I kinda like the consistency.

Considering this is a Suggestion I may or may not just resolve this, maybe adding a comment to the code or something.

Reply:
I think this is covered above, but whichever it done now as far as I understand.